cluster-api/api/cluster/v1alpha1/types.go

```diff
@@ -301,3 +301,66 @@ type MachineList struct {
 	metav1.ListMeta `json:"metadata"`
 	Items           []Machine `json:"items"`
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type MachineSet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec   MachineSetSpec   `json:"spec"`
+	Status MachineSetStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type MachineSetList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []MachineSet `json:"items"`
+}
+
+// MachineSetSpec is the specification of a MachineSet.
+type MachineSetSpec struct {
+	// Replicas is the number of desired replicas.
+	// This is a pointer to distinguish between explicit zero and unspecified.
+	// Defaults to 1.
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty"`
+
+	// Selector is a label query over machines that should match the replica count.
+	// Label keys and values that must match in order to be controlled by this MachineSet.
+	// It must match the machine template's labels.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+	Selector *metav1.LabelSelector `json:"selector"`
+
+	// Template is the object that describes the machine that will be created if
+	// insufficient replicas are detected.
+	// +optional
+	Template MachineTemplateSpec `json:"template,omitempty"`
```

Comment:
Why do we have a `MachineTemplateSpec` that nests a `MachineSpec`? 

That seems somewhat redundant at first glass. Wondering what the engineering reasoning here was.

Reply:
@kris-nova this is how `ReplicaSet` and `Deployment` works. For the template you need both `MachineSpec` and `ObjectMetadata`. `ObjectMetadata` is needed in case you want to set some extra labels and annotations for every machine.